contracts/core/EntryPoint.sol

```diff
@@ -819,7 +822,7 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
 
     function _getUnusedGasPenalty(uint256 gasUsed, uint256 gasLimit) internal pure returns (uint256) {
         unchecked {
-            if (gasLimit <= gasUsed) {
+            if (gasLimit <= gasUsed || gasLimit - gasUsed <= PENALTY_GAS_THRESHOLD) {
```

Comment:
Damn solidity, doesn't optimize `gasLimit-gasUsed` which is done twice...
better add below `if (unusedGas<=PENALTY_GAS_THRESHOLD) return 0;`
(which is also slightly more readable)